src/pipdeptree/_models/package.py

```diff
@@ -118,31 +110,73 @@ class DistPackage(Package):
 
     """
 
-    def __init__(self, obj: DistInfoDistribution, req: ReqPackage | None = None) -> None:
-        super().__init__(obj)
+    def __init__(self, obj: Distribution, req: ReqPackage | None = None) -> None:
+        super().__init__(obj.metadata["Name"])
+        self._obj = obj
         self.req = req
-        self._project_name = ""
 
     def requires(self) -> list[Requirement]:
-        return self._obj.requires()  # type: ignore[no-untyped-call,no-any-return]
+        req_list = []
+        req_name_list = []
+        if self._obj.requires:
+            for r in self._obj.requires:
+                req = Requirement(r)
+                is_extra_req = req.marker and contains_extra(str(req.marker))
+                if not is_extra_req and req.name not in req_name_list:
+                    req_list.append(req)
+                    req_name_list.append(req.name)
+        return req_list
 
     @property
-    def project_name(self) -> str:
-        if not self._project_name:
-            try:
-                self._project_name = metadata(self.key)["name"]
-            except (PackageNotFoundError, KeyError):
-                self._project_name = self._obj.project_name
-        return self._project_name
+    def editable(self) -> bool:
+        return self.direct_url_dict["editable"]
+
+    @property
+    def direct_url(self) -> str:
+        return self.direct_url_dict["url"]
+
+    @property
+    def raw_name(self) -> str:
+        return self.project_name
 
     @property
     def version(self) -> str:
         return self._obj.version  # type: ignore[no-any-return]
 
+    @property
+    def editable_project_location(self) -> str:
+        if self.direct_url:
+            from pip._internal.utils.urls import url_to_path  # noqa: PLC2701, PLC0415
+
+            return url_to_path(self.direct_url)
+
+        egg_link_path = egg_link_path_from_sys_path(self.raw_name)
+        if egg_link_path:
+            return self.location
+        return None
```

Comment:
Think you meant to return egg_link_path? I need to do some research on what is going on here.

Also return "" not `None`

Reply:
It seems that `self.location` needs to be replaced which not exsit in DistPackage.

Reply:
After I run `python3 setup.py develop`,  the `.egg-link` is added to `site-packages` and the directory where source is in (e.g.  /data/python-munkres/munkres-1.1.4) is added to `sys.path`. The `/data/python-munkres/munkres-1.1.4` is the `editable_project_location` wants to return.
ref: https://github.com/pypa/pip/blob/f5e4ee104e7b171a7cfb2843c9c602abf7a4e346/src/pip/_internal/utils/egg_link.py#L33

So what we need to do is to read the `egg_link_path` (if exsits) and take out the source location.

Reply:
Ref to [here](https://github.com/pypa/pip/blob/f5e4ee104e7b171a7cfb2843c9c602abf7a4e346/src/pip/_internal/metadata/base.py#L331), `editable` is determined by `editable_project_location` , but it seems that our method is also right too.